File: samples/node/test/protocol.test.js
import assert from 'node:assert/strict';
import test from 'node:test';
import {
  HEADER_LENGTH,
  MAX_PACKET_LENGTH,
  MessageType,
  ProtocolError,
  decode,
  encode,
  frameLength,
} from '../src/protocol.js';

test('PHXP v1 messages round trip through the fixed envelope', () => {
  const connectionId = Buffer.alloc(16, 0xab);
  const messages = [
    { type: MessageType.HELLO },
    { type: MessageType.READY },
    {
      type: MessageType.HANDOFF,
      connectionId,
      peekedLength: 517,
      acceptedAtNs: 42n,
      requestedSni: 'www.contoso.com',
    },
    { type: MessageType.ADOPTED, connectionId },
    { type: MessageType.REJECTED, connectionId, rejectionCode: 7 },
  ];
  for (const message of messages) {
    const decoded = decode(encode(message));
    assert.equal(decoded.type, message.type);
    assert.deepEqual(decoded.connectionId, message.connectionId ?? Buffer.alloc(16));
    assert.equal(decoded.peekedLength, message.peekedLength ?? 0);
    assert.equal(decoded.acceptedAtNs, message.acceptedAtNs ?? 0n);
    assert.equal(decoded.requestedSni, message.requestedSni ?? '');
    assert.equal(decoded.rejectionCode, message.rejectionCode ?? 0);
  }
});

test('malformed and oversized PHXP frames are rejected', () => {
  const hello = encode({ type: MessageType.HELLO });
  assert.equal(hello.length, HEADER_LENGTH);
  assert.throws(() => decode(hello.subarray(0, 39)), ProtocolError);

  const badMagic = Buffer.from(hello);
  badMagic[0] = 0;
  assert.throws(() => decode(badMagic), /invalid magic/);

  const badVersion = Buffer.from(hello);
  badVersion[4] = 2;
  assert.throws(() => decode(badVersion), /unsupported/);

  const oversized = Buffer.from(hello);
  oversized.writeUInt16BE(MAX_PACKET_LENGTH - HEADER_LENGTH + 1, 36);
  assert.throws(() => frameLength(oversized), /exceeds/);

  assert.throws(
    () => encode({ type: MessageType.REJECTED, connectionId: Buffer.alloc(16) }),
    /unexpected/,
  );
});

File: samples/go/phxp/protocol_test.go
package phxp

import (
	"bytes"
	"encoding/binary"
	"testing"
)

func TestProtocolRoundTrip(t *testing.T) {
	id := [16]byte{0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab, 0xab}
	messages := []Message{
		{Type: TypeHello},
		{Type: TypeReady},
		{
			Type:         TypeHandoff,
			ConnectionID: id,
			PeekedLength: 517,
			AcceptedAtNS: 42,
			RequestedSNI: "www.contoso.com",
		},
		{Type: TypeAdopted, ConnectionID: id},
		{Type: TypeRejected, ConnectionID: id, RejectionCode: 7},
	}
	for _, message := range messages {
		packet, err := Encode(message)
		if err != nil {
			t.Fatalf("Encode(%v): %v", message.Type, err)
		}
		decoded, err := Decode(packet)
		if err != nil {
			t.Fatalf("Decode(%v): %v", message.Type, err)
		}
		if decoded != message {
			t.Fatalf("round trip mismatch:\n got %#v\nwant %#v", decoded, message)
		}
	}
}

func TestProtocolRejectsMalformedEnvelopes(t *testing.T) {
	hello := mustEncode(Message{Type: TypeHello})
	cases := [][]byte{
		hello[:HeaderLength-1],
		append([]byte("X"), hello[1:]...),
		func() []byte {
			packet := append([]byte(nil), hello...)
			packet[4] = 2
			return packet
		}(),
		func() []byte {
			packet := append([]byte(nil), hello...)
			packet[6] = 1
			return packet
		}(),
		func() []byte {
			packet := append([]byte(nil), hello...)
			binary.BigEndian.PutUint16(packet[36:38], MaxPacketLength-HeaderLength+1)
			return packet
		}(),
		func() []byte {
			packet := append([]byte(nil), hello...)
			packet[8] = 1
			return packet
		}(),
	}
	for index, packet := range cases {
		if _, err := Decode(packet); err == nil {
			t.Fatalf("malformed case %d was accepted", index)
		}
	}
}

func TestStreamFramingIsBoundedAndExact(t *testing.T) {
	packet := mustEncode(Message{
		Type:         TypeHandoff,
		ConnectionID: [16]byte{1},
		RequestedSNI: "example.test",
		PeekedLength: 123,
		AcceptedAtNS: 456,
	})
	for boundary := 1; boundary < len(packet); boundary++ {
		frame, err := readStreamFrame(bytes.NewReader(packet[boundary:]), packet[:boundary])
		if err != nil {
			t.Fatalf("split %d: %v", boundary, err)
		}
		if !bytes.Equal(frame, packet) {
			t.Fatalf("split %d changed frame", boundary)
		}
	}

	coalesced := append(append([]byte(nil), mustEncode(Message{Type: TypeHello})...), mustEncode(Message{Type: TypeReady})...)
	if _, err := readStreamFrame(bytes.NewReader(nil), coalesced); err == nil {
		t.Fatal("coalesced frames were accepted")
	}
}
